containers/src/Kokkos_DynRankView.hpp

```diff
@@ -1749,9 +1749,13 @@ KOKKOS_INLINE_FUNCTION constexpr auto DynRankView<D, P...>::layout() const ->
     case 6: return Impl::as_view_of_rank_n<6>(*this).layout();
     case 7: return Impl::as_view_of_rank_n<7>(*this).layout();
     default:
-      Kokkos::Impl::throw_runtime_exception(
-          "Calling DynRankView::layout on DRV of unexpected rank " +
-          std::to_string(rank()));
+      KOKKOS_IF_ON_HOST(
+          Kokkos::Impl::throw_runtime_exception(
+              "Calling DynRankView::layout on DRV of unexpected rank " +
+              std::to_string(rank()));)
+      KOKKOS_IF_ON_DEVICE(
+          Kokkos::abort(
+              "Calling DYnRankView::layout on DRV of unexpected rank");)
```

Comment:
The OpenMPTarget build is unhappy with this:

```
In file included from /var/jenkins/workspace/Kokkos/build/containers/unit_tests/serial/TestSerial_DynViewAPI_generic.cpp:2:

In file included from /var/jenkins/workspace/Kokkos/containers/unit_tests/TestDynViewAPI_generic.hpp:45:

In file included from /var/jenkins/workspace/Kokkos/containers/unit_tests/TestDynViewAPI.hpp:50:

/var/jenkins/workspace/Kokkos/containers/unit_tests/../src/Kokkos_DynRankView.hpp:1760:1: error: non-void function does not return a value in all control paths [-Werror,-Wreturn-type]

}

^

/var/jenkins/workspace/Kokkos/containers/unit_tests/../src/Kokkos_DynRankView.hpp:2005:58: note: in instantiation of member function 'Kokkos::DynRankView<double, Kokkos::Serial>::layout' requested here

  return dst_type(prop_copy, Impl::reconstructLayout(src.layout(), src.rank()));

                                                         ^

/var/jenkins/workspace/Kokkos/containers/unit_tests/../src/Kokkos_DynRankView.hpp:2049:16: note: in instantiation of function template specialization 'Kokkos::Impl::create_mirror<double, Kokkos::Serial>' requested here

  return Impl::create_mirror(src, Kokkos::Impl::ViewCtorProp<>{});

               ^

/var/jenkins/workspace/Kokkos/containers/unit_tests/TestDynViewAPI.hpp:1166:18: note: in instantiation of function template specialization 'Kokkos::create_mirror<double, Kokkos::Serial>' requested here

    hx = Kokkos::create_mirror(dx);

                 ^

/var/jenkins/workspace/Kokkos/containers/unit_tests/TestDynViewAPI.hpp:711:5: note: in instantiation of member function 'Test::TestDynViewAPI<double, Kokkos::Serial>::run_test_scalar' requested here

    run_test_scalar();

    ^

/var/jenkins/workspace/Kokkos/containers/unit_tests/TestDynViewAPI_generic.hpp:48:43: note: in instantiation of member function 'Test::TestDynViewAPI<double, Kokkos::Serial>::run_tests' requested here

  TestDynViewAPI<double, TEST_EXECSPACE>::run_tests();

                                          ^
```

Reply:
SYCL has the same complaint

Reply:
```suggestion
              "Calling DYnRankView::layout on DRV of unexpected rank");)
              return {};
```
Might be a workable hack

Reply:
The similar error-handling fix worked in #5248 because that had the intended `return` unconditionally at the end of the function

Reply:
bloody hell, gonna try that